### ISSUE TITLE
weird ^M at the end of line on Windows?

### ISSUE THREAD

**Issue by neilfranci** [nvim-pack/nvim-spectre#207]
### I'm using Windows, and EOL is LF (yes, I am not a fan of CRLF)
**But whatever I replace text, it comes up with ^M at the end of the line.**

![image](https://github.com/nvim-pack/nvim-spectre/assets/66520542/5a92984e-6109-49dc-9a92-61e864d16a1b)

Default setup
```lua
return {
    "nvim-pack/nvim-spectre",
    dependencies = {
        "nvim-lua/plenary.nvim",
   },
   keys = {
    {
      "<leader>S",
      function()
        require("spectre").toggle()
      end,
      desc = "Spectre",
    },
  },
  opts = {},
}
```

[My config](https://github.com/neilfranci/nvim-config)

**Comment by EpsilonKu** [comment]
definitely `sed` issue. Seems like it converts files from LF to CRLF after replace. I dunno how to solve it. Seems like you need to run by hand to remove ^M. Will  be good if @windwp can help. 
I will try to reproduce issue.

**Comment by GitMurf** [comment]
I am having same issue and the biggest problem is that it then messes up my undo history as it looks like the entire file was replaced/changed because every line adds the ^M carriage return. Anyone make any progress on this or have any ideas?

**Comment by GitMurf** [comment]
It seems we need to be able to run `sed` with the `binary` flag maybe? Is there a way to pass configuration / flags to sed from spectre?

https://www.gnu.org/software/sed/manual/sed.html

![image](https://github.com/nvim-pack/nvim-spectre/assets/64155612/f091e3e0-4a35-4539-b69d-06cc4b1baf8c)

**Comment by EpsilonKu** [comment]
Yeah, sure.
```lua
replace_engine={
      ['sed']={
          cmd = "sed",
          args = {'--binary'},

      },
  },
```

**Comment by GitMurf** [comment]
yeah unfortunately that doesn't fix it :( it is some weird thing with `sed` and windows where it is saving the file in windows native format of crlf even when the file itself starts as lf unix style.

**Comment by GitMurf** [comment]
@EpsilonKu is `sd` option supposed to be officially supported now to use? i see mention in the readme but not much more and looks like maybe a recent PR? I am getting an error on Windows:

`E5108: Error executing lua ...kstart-data/lazy/nvim-spectre/lua/spectre/replace/sd.lua:19: attempt to index local 'file' (a nil value)`

I would not be surprised if it is a windows path issue where the file arg is not being properly escaped or something. `sed` works fine (other than the ^M \r line feed issue).

**Comment by EpsilonKu** [comment]
Yeah, `sd` was added recently. Hope @Zananok can help. Sadly I don't have ready windows to reproduce issue.

**Comment by Zananok** [comment]
Problem is the line above, which opens the file:
`    local file = io.open(value.filename, "r")`
So, it means its not finding the file. I can do a simple PR later to improve readability of the error, but if you add a `print(value.filename)` just above, you will probably see a non proper formulated path due to windows or something.

As for ^M, its windows tools that causes this. You can run "dos2unix filename" on the file and should fix it. Additionally, but I haven't tested, after you save the file with ^M, run on the nvim: `:!dos2unix`

**Comment by GitMurf** [comment]
Bot is marking as stale. Bumping to see if this is something someone is willing considering to fix down the road. Thanks!

**Comment by Anakael** [comment]
> Problem is the line above, which opens the file: ` local file = io.open(value.filename, "r")` So, it means its not finding the file. I can do a simple PR later to improve readability of the error, but if you add a `print(value.filename)` just above, you will probably see a non proper formulated path due to windows or something.
> 
> As for ^M, its windows tools that causes this. You can run "dos2unix filename" on the file and should fix it. Additionally, but I haven't tested, after you save the file with ^M, run on the nvim: `:!dos2unix`. But this tool is for linux, so I am not sure if you have access to it.
> 
> I run my nvim on wsl, to avoid these troublesome issues, but even so, mixing wsl with windows gave me a fair share of issues.

-> you will probably see a non proper formulated path due to windows or something.
Yes, that's because next check:

![Image](https://github.com/user-attachments/assets/6b166ccd-98b7-4b84-929a-d995ba2778e5)

On Windows path do no starts with `/`.

**Comment by Polynominal** [comment]
I am on windows too and it worked a few updates back, but having updated I had to make a few edits.

frankly pretty pathetic to rely on regex pattern matching to decide what is and isnt an absolute path when pleanary already provides a faculty for it. 
on sd.lua
```lua
sd.replace = function(self, value)
    ...
    local ppath = Path:new(value.filename)
    if not ppath:is_absolute() then
        value.filename = Path:new(cwd):joinpath(value.filename):absolute()
    end
````

the sd command would print out just the echo with '...' without actually changing it, I think it had something to do with the '
here is a working one instead:

continuing on sd.lua
```lua
sd.replace = function(self, value)
       ....
        -- Use `io.popen` to get the transformed line using `sd`
        local command = string.format([[echo "%s" | sd "%s" "%s"]],
        lines[value.lnum]:gsub('"','""'), 
        value.search_text:gsub('"','""'),
        value.replace_text:gsub('"','""'))

        local handle = io.popen(command, 'r')
        if handle then
            local transformedLine = handle:read('*a')
            transformedLine = transformedLine:gsub('\n$', '')
            transformedLine = transformedLine:gsub('""','"')
            transformedLine = transformedLine:sub(2,-3)
            handle:close()
            -- Replace the line in memory
            lines[value.lnum] = transformedLine -- Remove trailing newline added by `echo`
```

Unfortunately you cannot handle ./ inside of sd.lua as thats appended to a path eg "C:\Project\.\Project\file.txt"  so it needs to be striped out at the moment of reception.

on init.lua:537
```lua
M.search_handler = function()
  ...
          on_result = function(item)
            if not state.is_running then
                return
            end
            item.replace_text = ''
            if string.match(item.filename, '^%.%/') then
                item.filename = item.filename:sub(3, #item.filename)
            end
            item.filename = vim.fs.normalize(item.filename) -- this is the critical one
```

the vim.fs.normalize should be used directly after any filename is input

The wrong filename handling propagates throughout nvim causing issues with LSP's aswell and in case of following symbolic links can break everything
not to mention the swapfile madness that comes from it

I might be doing it too late as I am unaware where exactly item.filename comes from eg if we have to adjust each search system individually
another entry I found is on utils.lua:10 and that appears to have fixed my issues and I am finally able to search and replace
```lua
M.parse_line_grep = function(query)
    local t = { text = query }
    local _, _, filename, lnum, col, text = string.find(t.text, _regex_file_line)

    if filename == nil then
        return nil
    end
    local ok
    ok, lnum = pcall(tonumber, lnum)
    if not ok then
        return nil
    end
    ok, col = pcall(tonumber, col)
    if not ok then
        return nil
    end

    t.filename = vim.fs.normalize(filename) -- <- here
    t.lnum = lnum
    t.col = col
    t.text = text
    return t
end
````

I know we windows users are 2nd class citizens at best, but it would be nice if we were considered at least a little bit especially when tools for that are already available

Edit: I added " support, this likely only works on windows so needs to be if'd
